src/DynamoCoreWpf/ViewModels/Menu/PreferencesViewModel.cs

```diff
@@ -1314,6 +1316,78 @@ private void AddPythonEnginesOptions()
+            {
+                var ppath = packagePath == DynamoModel.BuiltInPackagesToken ? PathManager.BuiltinPackagesDirectory : packagePath;
+                packageBinPaths = packageBinPaths.Concat(Directory.GetFiles(ppath, "*.dll", SearchOption.AllDirectories)).ToList();
+            }
+
+            var dotNetRuntimePaths = Directory.GetFiles(RuntimeEnvironment.GetRuntimeDirectory(), "*.dll");
+            var dynCorePaths = new FileInfo(Assembly.GetExecutingAssembly().Location).Directory.GetFiles("*.dll", SearchOption.AllDirectories).Select(x => x.FullName);
+            var standardPaths = dotNetRuntimePaths.Concat(dynCorePaths);
+
+            MetadataLoadContext mlc = null;
+            var resolver = new PathAssemblyResolver(standardPaths);
+            mlc = new MetadataLoadContext(resolver);
+            var all = AppDomain.CurrentDomain.GetAssemblies().ToList();
+
+            foreach (var packageBin in packageBinPaths)
+            {
+                try
+                {
+                    //Load assembly into MLC
+                    var mlcassem = mlc.LoadFromAssemblyPath(packageBin);
+                    if (mlcassem != null)
+                    {
+                        //Check if it consists of a Python Engine
+                        var a = mlcassem.GetTypes().FirstOrDefault(x => IsEqualPythonEngineAssemblyName(x) && !x.IsInterface && !x.IsAbstract);
+                        if (a != null)
+                        {
+                            //Fetch it's loaded instance from AppDomain
+                            var b = all.FirstOrDefault(x => x.GetName().Name.ToLower().Equals(Path.GetFileNameWithoutExtension(mlcassem.GetName().Name).ToLower()), null);
+                            if (b != null)
+                            {
+                                //Validate result and add to map
+                                PythonEngineManager.Instance.ValidatePythonEngine(b, out var eng);
+                                if (PythonEnginesList.Contains(eng.Name))
+                                {
+                                    var packageLoader = dynamoViewModel.Model.GetPackageManagerExtension()?.PackageLoader;
+                                    var pkg = packageLoader.GetOwnerPackage(packageBin);
+                                    var tup = new Tuple<string, string>(pkg.Name, pkg.VersionName);
+                                    PythonEngineManager.Instance.PythonEnginePackageMap.Add(eng.Name, tup);
+                                }
+                            }
+                        }
+                    }
+                }
+                catch
+                {
+                    // Ignore exceptions from iterating assembly types.
+                    continue;
+                }
+            }
+            mlc?.Dispose();
+        }
+
+        //ignoring assembly version
+        private bool IsEqualPythonEngineAssemblyName(Type t)
```

Comment:
you found that this failed if the base types were of different versions? You cannot just check that the type is assignable to a `PythonEngine`?

Reply:
The assembly returned from MLC is different from the assembly loaded into the App, it had partial information, and the isAssignable check to `PythonEngine` type does not resolve (even though I can see the same types), that is why I had to go for this comparison.
Somewhat related: https://stackoverflow.com/questions/10439668/getting-custom-assembly-attributes-without-loading-into-current-appdomain